A86-xds-http-connect.md

```diff
@@ -0,0 +1,107 @@
+* Status: {Draft, In Review, Ready for Implementation, Implemented}
+* Implemented in: <language, ...>
+* Last updated: 2024-09-25
+* Discussion at: https://groups.google.com/g/grpc-io/c/WiqQ7h003fE
+
+## Abstract
+
+This document provides a design for gRPC to support configuring HTTP
+CONNECT functionality via xDS.
+
+## Background
+
+gRPC currently supports HTTP CONNECT proxies, as described in [gRFC A1].
+gRPC also supports obtaining configuration via xDS, as originally
+described in [gRFC A27].  This document describes new functionality to
+allow configuring use of an HTTP CONNECT proxy via xDS.
+
+### Related Proposals: 
+* [gRFC A1: HTTP CONNECT Proxy Support][gRFC A1]
+* [gRFC A27: xDS-Based Global Load Balancing][gRFC A27]
+* [gRFC A29: xDS-Based Security for gRPC Clients and Servers][gRFC A29]
+* [gRFC A83: A83: xDS GCP Authentication Filter][gRFC A83]
+
+[gRFC A1]: A1-http-connect-proxy-support.md
+[gRFC A27]: A27-xds-global-load-balancing.md
+[gRFC A29]: A29-xds-tls-security.md
+[gRFC A83]: A83-xds-gcp-authn-filter.md
+
+## Proposal
+
+The xDS configuration for HTTP CONNECT is done via the
+[`Http11ProxyUpstreamTransport`](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/extensions/transport_sockets/http_11_proxy/v3/upstream_http_11_connect.proto#L36)
+transport socket wrapper.  This wrapper is sent in the CDS [`transport_socket`
+field](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/config/cluster/v3/cluster.proto#L1099),
+as a wrapper around whatever transport socket is actually used.
+
+Note that the only transport socket implementation that gRPC currently
+supports is `UpstreamTlsContext`, which is used to configure TLS
+security, as per [gRFC A29].  We honor the xDS TLS configuration only if
+the application uses `XdsCredentials` when creating the gRPC channel.
+With this design, we will now honor the `Http11ProxyUpstreamTransport`
+transport socket wrapper regardless of whether `XdsCredentials` is used.
+However, we will retain the original behavior for the underlying
+transport socket wrapped by the `Http11ProxyUpstreamTransport` wrapper:
+if the underlying transport socket is `UpstreamTlsContext`, then we will
+honor it only if `XdsCredentials` is used, and if it is any other type,
+we will not use it.
+
+The `Http11ProxyUpstreamTransport` transport socket
+wrapper looks for the proxy address in the EDS metadata,
+both at the individual endpoint level (in the [`LbEndpoint.metadata`
+field](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/config/endpoint/v3/endpoint_components.proto#L122))
+and at the locality level (in the [`LocalityLbEndpoints.metadata`
+field](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/config/endpoint/v3/endpoint_components.proto#L165)).
+
+To support this, we will use the metadata mechanism described for CDS in
```

Comment:
I don't know what this means. This is different metadata than CDS. So we need to say what the "mechanism" is.

Reply:
It's actually not different metadata.  In Envoy, there's a single [`envoy.config.core.v3.Metadata`](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/config/core/v3/base.proto#L239) type that is used in a bunch of places -- including both the CDS metadata field described in A83 and the two EDS metadata fields described here.  Envoy has a single registry of supported protobuf types for metadata that it uses everywhere, so the set of values it accepts is the same in any field of this type, regardless of which resource type that field is in.

It's true that it wouldn't actually be useful to store the GCP auth audience in EDS metadata, nor would it be useful to store the proxy address in CDS metadata -- but that's because the things that looks for those metadata values is looking for them in a different metadata field, not because Envoy would reject the metadata if it was present in the wrong place.

I've attempted to clarify the wording to indicate that it's the metadata registry mechanism that we're reusing here.